### PR TITLE
Add more justification

### PR DESCRIPTION
**Pull request by chris-wood**
This also sharpens the requirement for the MAC, stating that it must be collision resistant. (We missed this before.)

cc @hugokraw

### CODE REVIEW

File: draft-irtf-cfrg-opaque.md [hunk 1]
```diff
@@ -1687,14 +1695,71 @@ protocols such as TLS.
...
+  Second, it removes the need for client applications to generate authentication
+  keys. Instead, this responsibility is handled by OPAQUE, thereby simplifying
+  the client interface to the protocol.
+- Envelopes are masked with a per-user masking key as a way of preventing
+  client enumeration attacks. See {{preventing-client-enumeration}} for more
+  details. This extension does not add to the security of OPAQUE as an aPAKE
+  but only used to provide a defense against enumeration attacks. In the
+  analysis, this key can be simulated as a (pseudo) random key. This change
+  was made to support real-world use cases where client or user enumeration
+  is a security (or privacy) threat.
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1725]:
```suggestion
  is a security (or privacy) risk.
```

File: draft-irtf-cfrg-opaque.md [hunk 2]
```diff
@@ -1687,14 +1695,71 @@ protocols such as TLS.
...
+- Per-user OPRF keys are derived from a client identity and cross-user seed
+  as a mitigation against client enumeration attacks. See
+  {{preventing-client-enumeration}} for more details. The analysis of OPAQUE
+  assumes OPRF keys of different users are independently random or
+  pseudorandom. Deriving these keys via a single PRF (i.e., with a single
+  cross-user key) applied to users' identities satisfies this assumption.
+  This change was made to support real-world use cases where client or user
+  enumeration is a security (or privacy) threat.
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1733]:
```suggestion
  enumeration is a security (or privacy) risk.
```

File: draft-irtf-cfrg-opaque.md [hunk 3]
```diff
@@ -1687,14 +1695,71 @@ protocols such as TLS.
...
+- The protocol outputs an export key for the client in addition to shared
+  session key that can be used for application-specific purposes. This key
+  is a pseudorandom value independent of other values in the protocol and
+  have no influence in the security analysis (it can be simulated with a
```
Comment (bytemare) [inline draft-irtf-cfrg-opaque.md:1737]:
```suggestion
  has no influence in the security analysis (it can be simulated with a
```

**Review by bytemare — CHANGES_REQUESTED** [review]

**Review by bytemare — APPROVED** [review]

File: draft-irtf-cfrg-opaque.md [hunk 4]
```diff
@@ -1684,19 +1700,121 @@ protocols such as TLS.
...
+  instantiates the OPRF functionality based on the {{I-D.irtf-cfrg-voprf}}, which
+  is identical to the DH-OPRF functionality in {{OPAQUE}} and, concretely, uses
+  the hash-to-curve functions in {{?I-D.irtf-cfrg-hash-to-curve}}. All hash-to-curve
+  methods in {{I-D.irtf-cfrg-hash-to-curve}} are compliant with the requirement
+  in {{OPAQUE}}, namely, that the output be a member of the prime-order group.
+- {{OPAQUE}} and {{I-D.krawczyk-cfrg-opaque-03}} both used HMQV as the AKE
+  for the protocol. However, this document fully specifies 3DH instead of HMQV
+  (though a sketch for how to instantiate OPAQUE using HMQV is included in {{hmqv-sketch}}).
+  Since 3DH satisfies the essential requirements for the AKE as described in {{OPAQUE}}
+  and {{I-D.krawczyk-cfrg-opaque-03}}, as described in {{security-analysis}}, this change
```
Comment (hugokraw) [inline draft-irtf-cfrg-opaque.md:1805]: change 'as described in {{security-analysis}}' with 'as recalled in {{security-analysis}}'

File: draft-irtf-cfrg-opaque.md [hunk 5]
```diff
@@ -1684,19 +1700,121 @@ protocols such as TLS.
...
+  has no effect on the overall security of the protocol. 3DH was chosen for its
```
Comment (hugokraw) [inline draft-irtf-cfrg-opaque.md:1806]: Change 'has not effect on" to "preserves"

File: draft-irtf-cfrg-opaque.md [hunk 6]
```diff
@@ -1684,19 +1700,121 @@ protocols such as TLS.
...
+  simplicity and ease of implementation.
+- The DH-OPRF and HMQV instantiation of OPAQUE in {{OPAQUE}}, Figure 12 uses
+  a different transcript than that which is described in this specification. In particular,
+  the key exchange transcript specified in {{ake-protocol}} is a superset of the transcript
+  as defined in {{OPAQUE}}. This was done to align with best practices, such as is
+  done for key exchange protocols like TLS 1.3 {{RFC8446}}.
+- Neither {{OPAQUE}} nor {{I-D.krawczyk-cfrg-opaque-03}} included wire format details for the
+  protocol, which is essential for interoperability. This specification rectifies this
```
Comment (hugokraw) [inline draft-irtf-cfrg-opaque.md:1814]: Replace 'rectifies' with 'fills'
(I was very explicit about these essential details not being in scope in my i-d, so I object to 'rectify' :-)

Comment (chris-wood) [reply draft-irtf-cfrg-opaque.md:1814]: Hah, fair :)

**Pull request merged**